File: react-native/src/config/configLoader.ts
import { getDynamicConfig } from '../native/measureBridge';
import type { Logger } from '../utils/logger';
import { DynamicConfig } from './dynamicConfig';

export class ConfigLoader {
  constructor(private logger: Logger) {}

  async loadDynamicConfig(): Promise<DynamicConfig | null> {
    let json: Record<string, any> | null = null;

    try {
      json = await getDynamicConfig();
    } catch (e) {
      this.logger.log(
        'info',
        'ConfigLoader: Failed to load dynamic config',
        e
      );
      return null;
    }

    if (!json) {
      this.logger.log(
        'info',
        'ConfigLoader: No dynamic config found, using defaults'
      );
      return null;
    }

      this.logger.log(
        'info',
        'ConfigLoader: Dynamic config loaded successfully'
      );
      console.log('Loaded dynamic config:', json);

      return DynamicConfig.fromNative(json);
  }
}

File: flutter/packages/measure_flutter/lib/src/config/config_loader.dart
import 'dart:convert';

import 'package:measure_flutter/src/config/dynamic_config.dart';
import 'package:measure_flutter/src/logger/log_level.dart';
import 'package:measure_flutter/src/logger/logger.dart';
import 'package:measure_flutter/src/storage/file_storage.dart';

class ConfigLoader {
  final Logger logger;
  final FileStorage fileStorage;

  const ConfigLoader({
    required this.logger,
    required this.fileStorage,
  });

  Future<DynamicConfig?> loadDynamicConfig() async {
    final content = await fileStorage.getDynamicConfigPath();
    if (content == null) {
      logger.log(LogLevel.info, 'ConfigLoader: No dynamic config found, using defaults');
      return null;
    }

    try {
      final json = jsonDecode(content) as Map<String, dynamic>;
      logger.log(LogLevel.debug, "ConfigLoader: Dynamic config loaded successfully");
      return DynamicConfig.fromJson(json);
    } catch (e, stacktrace) {
      logger.log(LogLevel.error, 'ConfigLoader: Error parsing dynamic config, using defaults', e, stacktrace);
      return null;
    }
  }
}
